apiml/src/main/java/org/zowe/apiml/filter/BasicLoginFilter.java

```diff
@@ -0,0 +1,141 @@
+        this.mapper = mapper;
+        this.failedAuthenticationWebHandler = failedAuthenticationWebHandler;
+    }
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+        String authHeader = exchange.getRequest().getHeaders().getFirst(HttpHeaders.AUTHORIZATION);
+        if (authHeader != null && authHeader.toLowerCase().startsWith("bearer ")) {
+            return chain.filter(exchange);
+        }
+        return extractBasicAuth(exchange)
+            .switchIfEmpty(getCredentialsFromBody(exchange))
+            .map(this::useCredentials)
+            .switchIfEmpty(chain.filter(exchange).then(Mono.empty()))
+            .flatMap(credentials ->
+                authenticationManager.authenticate(credentials)
+                    .flatMap(authentication -> chain.filter(exchange)
+                        .contextWrite(ReactiveSecurityContextHolder.withAuthentication(authentication))
+                    )
+                    .onErrorResume(AuthenticationException.class, ex -> failedAuthenticationWebHandler.onAuthenticationFailure(new WebFilterExchange(exchange, chain), ex))
+            ).onErrorResume(AuthenticationException.class, ex -> failedAuthenticationWebHandler.onAuthenticationFailure(new WebFilterExchange(exchange, chain), ex)
+            );
+    }
+
+    private AbstractAuthenticationToken useCredentials(LoginRequest credentials) {
+        return new UsernamePasswordAuthenticationToken(credentials.getUsername(), credentials.getPassword());
+    }
+
+    private Mono<LoginRequest> extractBasicAuth(ServerWebExchange exchange) {
+        var authHeader = exchange.getRequest().getHeaders().getFirst(HttpHeaders.AUTHORIZATION);
+        if (authHeader == null) return Mono.empty();
+        try {
+            var optHeader = LoginFilter.getCredentialFromAuthorizationHeader(Optional.of(authHeader));
+            return optHeader
+                .filter(this::credentialsAvailable)
+                .map(Mono::just)
+                .orElseGet(() -> Mono.error(new AuthenticationCredentialsNotFoundException("Username or password not provided.")));
+        } catch (Exception e) {
+            log.debug("Failed to decode Basic Auth header: {}", e.getMessage());
+            return Mono.error(new AuthenticationCredentialsNotFoundException("Invalid basic authentication header", e));
+        }
+
+    }
+
+    private boolean credentialsAvailable(LoginRequest credentials) {
+        return credentials.getUsername() != null && !credentials.getUsername().isBlank()
+            && credentials.getPassword() != null && credentials.getPassword().length > 0;
+    }
+
+    private Mono<LoginRequest> getCredentialsFromBody(ServerWebExchange exchange) {
+        var path = exchange.getRequest().getPath().value();
+        if (!path.contains("/auth/login")) {
+            return Mono.empty();
+        }
+        return exchange.getRequest().getBody().flatMap(buffer -> {
```

Comment:
I guess flatMap points to there could be multiple buffers. Therefore, this implementation is potentially wrong; the JSON could probably be split.

We should use an approach with `DataBufferUtils.join` to join buffers and then parse it (see ie. on https://medium.com/@jijith.srk/read-request-body-in-spring-webflux-webfilter-bddf2564b92d).

Reply:
done